src/Utilities/ExpressionExtension.cs

```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Linq.Expressions;
+
+namespace ePiggyWeb.Utilities
+{
+    public static class ExpressionExtension
+    {
+        public static Expression<Func<T, bool>> Or<T>(this Expression<Func<T, bool>> expression1, Expression<Func<T, bool>> expression2)
+        {
+            return expression1.Compose(Expression.Or, expression2);
+        }
+
+        public static Expression<Func<T, bool>> And<T>(this Expression<Func<T, bool>> expression1, Expression<Func<T, bool>> expression2)
+        {
+            return expression1.Compose(Expression.And, expression2);
+        }
+
+        private static Expression<Func<T, bool>> Compose<T>(this Expression<Func<T, bool>> expression1,
+            Func<Expression, Expression, BinaryExpression> logicalOperation, Expression<Func<T, bool>> expression2)
+        {
+            var invokedExpression = Expression.Invoke(expression2, expression1.Parameters);
+            return Expression.Lambda<Func<T, bool>>(logicalOperation.Invoke(expression1.Body, invokedExpression), expression1.Parameters);
```

Comment:
Source to read or just the difference between these? well what is different between first and second I understand, but what about third?

Reply:
First thing to note, is what is the difference between Expression<Func<T1, T2>> and Func<T1,T2> is. The first is for simplicity sake, lets say a wrapped Func delegate, Expression allows building expression trees, as I understand, things like I'm making here. For example

     Expression<Func<int, bool>> expression = x => x < 2;
     Func<int, bool> function = expression.Compile();
(However, you can't create an Expression from a Func)

The compose method takes in the following parameters:

     this Expression<Func<T, bool>> expression1
     Expression<Func<T, bool>> expression2

are the two predicates we want to compose/fuse/merge into one.

    Func<Expression, Expression, BinaryExpression> logicalOperation

is a delegate, that takes two Expressions as parameters, and returns a Binary Expression (basically this is how I store the logical operation we want to merge by, like 'And' or 'Or').

            var invokedExpression = Expression.Invoke(expression2, expression1.Parameters);

this is where it gets a difficult to explain ( I don't understand this completely myself). Basically, invoking a Lambda Expression is applying an argument(s) ( in this case expression1 parameters) to a Lambda Expression ( in this case expression2). For example, invoking would look like

     Expression<Func<int, bool>> pred1 = x => x > 1000;
     Expression<Func<int, bool>> pred2 = y => y == 20;
     Expression.Invoke(pred1, pred1.Parameters)
     //Invoke(x => (x > 1000), y)      - this is how it looks printed out


            return Expression.Lambda<Func<T, bool>>(logicalOperation.Invoke(expression1.Body, invokedExpression), expression1.Parameters);

So to start, Expression.Lambda creates a Lambda Expression, taking the first parameter as a body of the Lambda Expression, and later on parameters as the parameters of the Lambda Expression.
Then we invoke the logical Operation delegate (note that this is different invoke from the previous one), basically meaning:

     logicalOperation.Invoke(param1, param2) == Expression.And(param1, param2)

and then we pass expression1 parameters to this Lambda Expression.

(Why we have to use invoked expression in the logical operation, I'm not sure)

Keep in mind parts of the code I got from the internet, now spent an hour or so trying to understand fully how it works, so the explanation might not be the most accurate.
Sources:
https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression.invoke?view=net-5.0
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions
https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.invocationexpression?view=net-5.0

Reply:
I tried my best to explain this, but I still it's not very clear what this does :D